Easier learning exercise for bitstrings - paint-by-numbers (#1295)
exercism/elixir: exercises/concept/paint-by-number/.docs/hints.md

# Hints

## General

- Read about [bitstrings][bitstring] in the Getting Started guide.
- Review the documentation for the [bitstring special form][bitstring-form].
- Review [how to convert decimal numbers to binary numbers]([decimal-to-binary-youtube]).
- Elixir supports [different ways of writing integers][integer-literal], including binary notation.

## 1. Calculate palette bit size

- This task doesn't use the bitstring data type.
- Find the smallest positive number `n` such that 2 raised to the power of `n` is greater than or equal than the number of colors we need to represent.

> - Create a recursive function for this.

Reviewer: I did not understand the instructions in th same way. I thought I was supposed to find a direct way.

Author: I assume that means you skipped reading the instructions midway and didn't see the sentence "Now, solve this task with recursion and..."?

- Use the [type operator][type-operator] to specify the bit size of the new fragment as well as the existing bitstring.
- Use the previously implemented `PaintByNumber.palette_bit_size/1` function to get the bit size of the new fragment.
- Use the special `::bitstring` type to specify that the old fragment is of unknown size.

## 5. Get the first pixel from a picture

- The [bitstring special form][bitstring-form] can be used to [pattern match bitstrings][bitstring-matching].
- Use the previously implemented `PaintByNumber.palette_bit_size/1` function to get the bit size of one pixel.
- Use the special `::bitstring` type to specify that the rest of the bitstring is of unknown size.

## 6. Drop the first pixel from a picture

- The [bitstring special form][bitstring-form] can be used to [pattern match bitstrings][bitstring-matching].
- Use the previously implemented `PaintByNumber.palette_bit_size/1` function to get the bit size of one pixel.
- Use the special `::bitstring` type to specify that the rest of the bitstring is of unknown size.

## 7. Concatenate two pictures

- The [bitstring special form][bitstring-form] can be used to concatenate two bitstrings.

> - Use the special `::bitstring` type to specify that the two bitstring fragments are of unknown size.

Reviewer: I was not confused by this, but the student might conclude that it is only possible to use the bitstring type once (for the last part of the construction process of a new bitstring).

Author: Would this tiny wording tweak help?
```suggestion
- Use the special `::bitstring` type to specify that each of the bitstring fragments is of unknown size.
```

Reviewer: Yes, I think that would show the difference to the previous use case better.
